plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+        rval = ns;
+    }
+
+    return rval;
+}
+
+/** Translate the structure to code.
+ *
+ * order is important, affects code layout:
+ *  - anonymouse for test double instance.
+ *  - implementations using double.
+ *  - adapter.
+ */
+void generate(CppRoot r, Controller ctrl, Parameters params,
+        Generator.Modules modules, ref const(Container) container)
+in {
+    import std.array : empty;
+
+    assert(r.funcRange.empty);
+}
+body {
+    import std.algorithm : each, filter;
+    import std.array;
+    import std.typecons;
+    import cpptooling.data.symbol.types : USRType;
+    import cpptooling.generator.func : generateFuncImpl;
+    import cpptooling.generator.gmock : generateGmock;
+    import cpptooling.generator.includes : generateIncludes;
+    import cpptooling.data.representation;
+    import cpptooling.analyzer.type;
+
+    generateIncludes(ctrl, params, modules.hdr);
+
+    static void gmockGlobal(T)(T r, CppModule gmock, Parameters params) {
+        foreach (a; r.filter!(a => cast(ClassType) a.kind == ClassType.Gmock)) {
+            generateGmock(a, gmock, params);
+        }
+    }
+
+    // recursive to handle nested namespaces.
+    // the singleton ns must be the first code generate or the impl can't
+    // use the instance.
+    @trusted static void eachNs(LookupT)(CppNamespace ns, Parameters params,
+            Generator.Modules modules, CppModule impl_singleton, LookupT lookup, ref string[] cifaces) {
+        import std.variant;
+        import std.algorithm : canFind;	
+	    string currnsrp;	
+        string currns = ns.fullyQualifiedName;
+
+        auto currns_spl = (cast(string)(currns)).split("::");
+	    bool isReqOrPro = currns_spl[$-1] == "Requirer" ||  currns_spl[$-1] == "Provider";
+        if (isReqOrPro)
+        {
+            currnsrp = join(currns_spl[0..$-1], "::");
+        }
+        else 
+        {
+            currnsrp = currns;
+        }
+        
+        SUTEnv sut = params.getSut.GetSUTFromNamespace(currnsrp);
+        auto inner = modules;
+        CppModule inner_impl_singleton;
+	    
+        if(sut.valid) {
+		
+		    final switch(cast(NamespaceType) ns.kind) with (NamespaceType) {
+	             case Normal:
+                         inner.hdr = modules.hdr.namespace(ns.name);
+                         inner.hdr.suppressIndent(1);
+                         inner.impl = modules.impl.namespace(ns.name);
+                         break;
+                     case TestDoubleSingleton:
+                         break;
+                     case TestDouble:
+                         break;
+                 }
+	    }
+        
+        if (sut.valid && isReqOrPro)
+        {
+            foreach (a; ns.funcRange) {
+                generateFuncImpl(a, inner.impl);
+            }
+
+            foreach (a; ns.classRange) {
+                foreach (b; a.methodPublicRange) { 
+                    b.visit!((const CppMethod a) => generateCppMeth(a, inner, ns.fullyQualifiedName, sut),
```

Comment:
The nested loops could be combined with joiner.
Example:
```D
ns.classRange.map!(a => a.methodPublicRange).joiner
```
(I think, comment coding isn't my strong side)